### PR TITLE
05 - added data validation and GlobalExceptionHandler

### PR DESCRIPTION
**Pull request by yevhenkukharchuk**

### CODE REVIEW

File: src/main/java/yevhen/bookstore/dto/CreateBookRequestDto.java [hunk 1]
```diff
@@ -1,11 +1,18 @@
 package yevhen.bookstore.dto;
 
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
 import java.math.BigDecimal;
 
 public record CreateBookRequestDto(
+        @NotNull
```
Comment (YuraNR) [inline src/main/java/yevhen/bookstore/dto/CreateBookRequestDto.java]: Will be better to use @NotBlank

File: src/main/java/yevhen/bookstore/dto/CreateBookRequestDto.java [hunk 2]
```diff
@@ -1,11 +1,18 @@
...
         String title,
+        @NotNull
         String author,
+        @NotNull
         String isbn,
+        @NotNull
+        @Min(0)
```
Comment (YuraNR) [inline src/main/java/yevhen/bookstore/dto/CreateBookRequestDto.java]: Also we need to use here @PositiveOrZero instead @Min(0)

File: src/main/java/yevhen/bookstore/exception/CustomGlobalExceptionHandler.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+public class CustomGlobalExceptionHandler extends ResponseEntityExceptionHandler {
+    @Override
+    protected ResponseEntity<Object> handleMethodArgumentNotValid(
+            MethodArgumentNotValidException ex,
+            HttpHeaders headers,
+            HttpStatusCode status,
+            WebRequest request) {
+        Map<String, Object> errorMessage = new LinkedHashMap<>();
+        errorMessage.put("timestamp", LocalDateTime.now());
+        errorMessage.put("status", HttpStatus.BAD_REQUEST);
```
Comment (YuraNR) [inline src/main/java/yevhen/bookstore/exception/CustomGlobalExceptionHandler.java]: errorMessage.put("timestamp", LocalDateTime.now());
errorMessage.put("status", HttpStatus.BAD_REQUEST);
we don't need it

**Review by YuraNR — CHANGES_REQUESTED** [review]

**Review by YuraNR — APPROVED** [review]

File: pom.xml [hunk 4]
```diff
@@ -67,6 +67,10 @@
             <artifactId>mapstruct</artifactId>
             <version>${mapstruct.version}</version>
         </dependency>
+        <dependency>
+            <groupId>org.hibernate.validator</groupId>
+            <artifactId>hibernate-validator</artifactId>
+        </dependency>
```
Comment (liliia-ponomarenko) [inline pom.xml]: it is better to use spring-boot-starter-validation

File: src/main/java/yevhen/bookstore/controller/BookController.java [hunk 5]
```diff
@@ -35,7 +36,7 @@ public BookDto getBookById(@PathVariable Long id) {
 
     @PostMapping
     @ResponseStatus(HttpStatus.CREATED)
-    public BookDto createBook(@RequestBody CreateBookRequestDto requestDto) {
+    public BookDto createBook(@RequestBody @Valid CreateBookRequestDto requestDto) {
```
Comment (liliia-ponomarenko) [inline src/main/java/yevhen/bookstore/controller/BookController.java:39]: You should  add Valid to the body of the update method

File: src/main/java/yevhen/bookstore/exception/CustomGlobalExceptionHandler.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+import org.springframework.http.ResponseEntity;
+import org.springframework.validation.FieldError;
+import org.springframework.validation.ObjectError;
+import org.springframework.web.bind.MethodArgumentNotValidException;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.context.request.WebRequest;
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
+
+@ControllerAdvice
+public class CustomGlobalExceptionHandler extends ResponseEntityExceptionHandler {
```
Comment (liliia-ponomarenko) [inline src/main/java/yevhen/bookstore/exception/CustomGlobalExceptionHandler.java:20]: You also should handle EntityNotFoundException

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s  improve your solution)

File: pom.xml [hunk 7]
```diff
@@ -67,6 +67,11 @@
             <artifactId>mapstruct</artifactId>
             <version>${mapstruct.version}</version>
         </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-validation</artifactId>
+            <version>3.1.2</version>
```
Comment (NazarSmith228) [inline pom.xml:73]: you don't need to specify version here - it is taken from `spring-boot-starter-parent`

File: src/main/java/yevhen/bookstore/dto/CreateBookRequestDto.java [hunk 8]
```diff
@@ -1,11 +1,18 @@
...
 import java.math.BigDecimal;
 
 public record CreateBookRequestDto(
+        @NotBlank
         String title,
+        @NotBlank
         String author,
+        @NotBlank
         String isbn,
+        @NotBlank
```
Comment (NazarSmith228) [inline src/main/java/yevhen/bookstore/dto/CreateBookRequestDto.java:14]: I guess you should use `@NotNull` with BigDecimal

**Review by NazarSmith228 — APPROVED** [review]
LGMT, but pls pay attention to my remarks

**Pull request merged**